### PR TITLE
BAPS3 Presenter Merge In

### PR DESCRIPTION
**Pull request by mstratford**
Now this is going to be a mess.

### CODE REVIEW

**Comment by mstratford** [comment]
Last known thing to be "fixed" is the webstudio logo doing a weird fade in thing, but is otherwise ready for scrutiny I think.

**Comment by mstratford** [comment]
That is fixed now, looks all good for going to webstudio-dev.

File: package.json [hunk 1]
```diff
@@ -111,10 +111,12 @@
     "workbox-webpack-plugin": "^5.1.2",
     "worklet-loader": "^1.0.0"
   },
-  "homepage": "https://ury.org.uk/webstudio",
+  "homepage": "./",
```
Comment (markspolakovs) [inline package.json:114]: This broke the automated Jenkins builds. @mstratford is this what it needs to be for baps3? If so it's fine, we'll just need to amend the jenkins scripts.

Comment (mstratford) [reply package.json:114]: hmm, will check, but I think it was when making a build, the links to resources would be all wrong.

Comment (markspolakovs) [reply package.json:114]: I’ve worked out a way to have it be more reliable - it’s live on my manual Jenkins job, will merge into the Jenkinsfile

**Comment by mstratford** [comment]
I will try and get this finished tomorrow.

**Comment by mstratford** [comment]
Bugs noticed:
- Set Played / Unplayed in BAPSicle doesn't work.

**Comment by mstratford** [comment]
Right, just working on implementing the playedAt timing stuff into BAPSicle, then we can merge.

**Comment by mstratford** [comment]
As far as I can tell, both WebStudio and BAPS Presenter are working as expected. Ready for merge.

File: src/App.tsx [hunk 2]
```diff
@@ -15,9 +18,29 @@ function useForceUpdate() {
...
+    currentTimeslot,
+    timeslotLoading,
+  } = useSelector((state: RootState) => state.session);
+
+  if (process.env.REACT_APP_BAPSICLE_INTERFACE) {
+    document.title = "BAPS3 Presenter";
+    if (bapsConnectionState !== "CONNECTED") {
+      return <BAPSSessionHandler />;
+    } else {
+      return <Showplanner timeslotId={-1} />;
```
Comment (markspolakovs) [inline src/App.tsx:40]: What's the significance of -1 here?

File: src/api.ts [hunk 3]
```diff
@@ -55,6 +55,20 @@ export async function myradioApiRequest(
   }
 }
 
+export async function bapsicleApiRequest(
+  endpoint: string,
+  method: "GET" | "POST" | "PUT",
+  params: any
+): Promise<any> {
+  const res = await apiRequest(
+    "http://" + window.location.hostname + ":13500" + endpoint,
```
Comment (markspolakovs) [inline src/api.ts:64]: Probably something for the future, but would be nice if this wasn't hardcoded.

File: src/bapsicle.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            store.dispatch(
+              MixerState.setRepeat({
+                player: channel,
+                mode: bapsicle_state.repeat.toLowerCase(),
+              })
+            );
+
+            if (!("show_plan" in message.data)) {
+              console.error("Show plan data missing from status");
+              console.error(message.data);
```
Comment (markspolakovs) [inline src/bapsicle.ts:107]: This will execute the next lines even if this is true. I don't know bapsicle's API well enough to know how likely this is, but that seems prone to runtime errors - should there be an `else` on this `if`?

File: src/bapsicle.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                timeslotitemid: String(raw_planitem.timeslotItemId),
+                channel: message.channel,
+                played: false,
+                playedAt: raw_planitem.played_at,
+                playCount: raw_planitem.play_count,
+                type: raw_planitem.trackId ? "central" : "aux",
+                trackid: raw_planitem.trackId,
+                managedid: raw_planitem.managedId,
+                auxid: raw_planitem.managedId,
+                ...raw_planitem,
```
Comment (markspolakovs) [inline src/bapsicle.ts:123]: This will overwrite any fields that have the same name as ones specified earlier. Is this intentional?

File: src/bapsicle.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+              };
+              planItem = planItem as PlanItem;
```
Comment (markspolakovs) [inline src/bapsicle.ts:125]: Why is this needed? We already do `var planItem: PlanItem`.

File: src/bapsicle.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+                channel: message.channel,
+                planItems: planItems,
+              })
+            );
+        }
+      } else if ("message" in message) {
+        if (message.message === "Hello") {
+          //store.dispatch(
+          //  connection.actions.setServerState(message.serverName)
+          //);
```
Comment (markspolakovs) [inline src/bapsicle.ts:139]: Can this be removed?

File: src/bapsicle.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+      } else {
+        console.log("Unhandled: ", message);
+      }
+    };
+  }
+  return next(action);
+};
+
+export function sendBAPSicleChannel(message: any): void {
+  if (BAPSicleWS) {
```
Comment (markspolakovs) [inline src/bapsicle.ts:150]: This will drop messages if this is called before the connection is established. I imagine it's fairly unlikely that this will happen, but it may be worth sticking in a `console.warn` - if nothing else, it'll make debugging easier if it does end up happening.

File: src/bapsiclesession/state.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+export default sessionState.reducer;
+
+export const { setServerState } = sessionState.actions;
+
+export const getCurrentServer = (): AppThunk => async (dispatch, getState) => {
+  return getState().bapsSession.currentServer;
+};
+
+export const getServer = (): AppThunk => async (dispatch) => {
+  // TODO Server Details Configurable
```
Comment (markspolakovs) [inline src/bapsiclesession/state.ts:55]: Reminder to file an issue for this, rather than have the TODO get lost in the codebase :p

File: src/mixer/state.ts [hunk 10]
```diff
@@ -998,74 +1175,129 @@ export const mixerMiddleware: Middleware<{}, RootState, Dispatch<any>> = (
...
-  Keys("j", () => {
-    store.dispatch(setVolume(2, "off"));
-  });
-  Keys("k", () => {
-    store.dispatch(setVolume(2, "bed"));
-  });
-  Keys("l", () => {
-    store.dispatch(setVolume(2, "full"));
-  });
+  if (!process.env.REACT_APP_BAPSICLE_INTERFACE) {
```
Comment (markspolakovs) [inline src/mixer/state.ts:1267]: Might it be worth keeping these around even in bapsicle mode? It'd make switching between webstudio and baps easier, as people don't have to relearn their muscle memory. Feel free to overrule me though.

File: src/showplanner/PISModal.tsx [hunk 11]
```diff
@@ -46,16 +46,8 @@ function PersistNotice() {
       <div className="p-2 alert-primary">
         <h2>Welcome to WebStudio {process.env.REACT_APP_VERSION}!</h2>
         <p>
-          As of this version,{" "}
```
Comment (markspolakovs) [inline src/showplanner/PISModal.tsx:49]: Should we just get rid of `PersistNotice` entirely?

**Review by markspolakovs — COMMENTED** [review]
Left comments inline - mostly nitpicking, I'm happy with it as a whole. Will run it locally in WS mode next.

**Comment by markspolakovs** [comment]
I had to wrestle a bit with `node-sass` and `sass-loader` to get a version set that works on node 16 (node-sass 6 and sass-loader 10 appear to work) but local WS doesn't appear to have broken with this. Good job Matt!

File: src/App.tsx (hunk 2 above)
Comment (mstratford) [reply src/App.tsx:40]: That it required a number type for WebStudio use :P Probably can be made optional.

File: src/api.ts (hunk 3 above)
Comment (mstratford) [reply src/api.ts:64]: Mmm, yeah. I'll have a go. When built inside BAPSicle, it's expected to be whatever the port the server is running on anyway, but in separate development it may want to be configurable in the env.

File: src/mixer/state.ts (hunk 10 above)
Comment (mstratford) [reply src/mixer/state.ts:1267]: hmm, it was to mainly get rid of the "i'm typing into the search box" bug where possible. Especially given at least the studio keyboards are labeled with BAPS keys.

File: src/showplanner/PISModal.tsx (hunk 11 above)
Comment (mstratford) [reply src/showplanner/PISModal.tsx:49]: Maybe!

File: src/mixer/state.ts (hunk 10 above)
Comment (markspolakovs) [reply src/mixer/state.ts:1267]: Hrm, I see what you're getting at, but I can also imagine a hypothetical future scenario where you have a bapsicle backend but browser frontend, where that won't work.

Eh, we can cross that bridge when we come to it.

File: src/bapsicle.ts (hunk 7 above)
Comment (mstratford) [reply src/bapsicle.ts]: Actually implemented showing the server name again instead.

File: src/bapsicle.ts (hunk 6 above)
Comment (mstratford) [reply src/bapsicle.ts]: Removed.

File: src/bapsicle.ts (hunk 5 above)
Comment (mstratford) [reply src/bapsicle.ts]: Tidied up this, quite a lot of things were no longer necessary even if they did work.

File: README.md [hunk 12]
```diff
@@ -30,15 +30,21 @@ $ source venv/bin/activate
...
+
+See below for how to work on each one.
+
+### Hacking (WebStudio)
+
+Start the client by running `npm start`.
 
 Start the server by running `python3 stateserver.py` and `python3 shittyserver.py`.
 
 Don't forget to ensure that both TypeScript and MyPy pass, as your code will be rejected by CI otherwise - run `tsc --noEmit` and/or `mypy server.py` to check.
```
Comment (markspolakovs) [inline README.md:45]: This line needs changing (more of a note to self than a ding of your PR)

File: src/bapsiclesession/state.ts [hunk 13]
```diff
@@ -45,19 +52,31 @@ const sessionState = createSlice({
...
+      : window.location.hostname,
+    ws_port: process.env.REACT_APP_WEBSOCKET_PORT
+      ? parseInt(process.env.REACT_APP_WEBSOCKET_PORT)
+      : 13501,
+    ui_protocol: process.env.REACT_APP_BAPSICLE_PROTOCOL
+      ? process.env.REACT_APP_BAPSICLE_PROTOCOL
+      : "http",
+    ui_port: process.env.REACT_APP_BAPSICLE_PORT
+      ? parseInt(process.env.REACT_APP_BAPSICLE_PORT)
+      : parseInt(window.location.port),
```
Comment (markspolakovs) [inline src/bapsiclesession/state.ts:75]: Pretty sure these can be simplified using `??` (e.g. `process.env.REACT_APP_BAPSICLE_HOST ?? window.location.hostname`)

**Review by markspolakovs — DISMISSED** [review]

**Comment by markspolakovs** [comment]
Note that I haven't re-tested it, so I'm assuming you haven't broken the build in the meantime ;)

File: src/bapsiclesession/state.ts (hunk 13 above)
Comment (mstratford) [reply src/bapsiclesession/state.ts]: Cheers :) Done.

**Comment by mstratford** [comment]
I've rebuilt both and they seem good :)

File: src/bapsiclesession/state.ts [hunk 14]
```diff
@@ -4,11 +4,11 @@ import { Timeslot } from "../api";
...
-  hostname: String | null;
-  ui_port: Number | null;
-  ui_protocol: String | null;
-  ws_port: Number | null;
-  name: String | null;
+  hostname: String;
+  ui_port: Number;
+  ui_protocol: String;
+  ws_port: Number;
+  name: String;
```
Comment (markspolakovs) [inline src/bapsiclesession/state.ts:11]: Should be `string` and `number` (lower-case) ;)
